File: server/datasets/java_source-code/Multiple-C2/organic_9.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #9
 * Verbose style with basic input validation.
 */

class Solution {
    public static int knapsackSolver(int[] itemWeights, int[] itemValues, int maxCapacity) {
        if (itemWeights.length != itemValues.length) {
            throw new IllegalArgumentException("weights and values must be same length");
        }
        if (maxCapacity < 0) {
            throw new IllegalArgumentException("capacity cannot be negative");
        }

        int numItems = itemWeights.length;
        int[][] grid = new int[numItems + 1][maxCapacity + 1];

        for (int itemIndex = 1; itemIndex <= numItems; itemIndex++) {
            int currentWeight = itemWeights[itemIndex - 1];
            int currentValue = itemValues[itemIndex - 1];

            for (int currentCapacity = 0; currentCapacity <= maxCapacity; currentCapacity++) {
                if (currentWeight > currentCapacity) {
                    grid[itemIndex][currentCapacity] = grid[itemIndex - 1][currentCapacity];
                } else {
                    int valueIfSkipped = grid[itemIndex - 1][currentCapacity];
                    int valueIfTaken = currentValue + grid[itemIndex - 1][currentCapacity - currentWeight];
                    grid[itemIndex][currentCapacity] = Math.max(valueIfSkipped, valueIfTaken);
                }
            }
        }
        return grid[numItems][maxCapacity];
    }

    public static void main(String[] args) {
        int[] weights = {1, 2, 3, 8, 7, 4};
        int[] values = {20, 5, 10, 40, 15, 25};
        int capacity = 10;
        int best = knapsackSolver(weights, values, capacity);
        System.out.println("Best possible value: " + best);
    }
}

File: server/datasets/controlled_benchmarks/python/multiple_files/multiple_c2/organic_9.py
"""
0/1 Knapsack (DP) - Organic Submission #9
Verbose, heavily-commented beginner style with basic input validation.
"""


def knapsack_solver(item_weights, item_values, max_capacity):
    # basic validation
    if len(item_weights) != len(item_values):
        raise ValueError("weights and values must be the same length")
    if max_capacity < 0:
        raise ValueError("capacity cannot be negative")

    num_items = len(item_weights)

    # create a (num_items+1) x (max_capacity+1) grid filled with zeros
    grid = []
    for row in range(num_items + 1):
        grid.append([0 for _ in range(max_capacity + 1)])

    # fill the grid row by row
    for item_index in range(1, num_items + 1):
        current_weight = item_weights[item_index - 1]
        current_value = item_values[item_index - 1]

        for current_capacity in range(max_capacity + 1):
            # can we even fit this item?
            if current_weight > current_capacity:
                # nope, so the best we can do is skip it
                grid[item_index][current_capacity] = grid[item_index - 1][current_capacity]
            else:
                # we can choose to take it or leave it - pick the better option
                value_if_skipped = grid[item_index - 1][current_capacity]
                value_if_taken = current_value + grid[item_index - 1][current_capacity - current_weight]
                grid[item_index][current_capacity] = max(value_if_skipped, value_if_taken)

    return grid[num_items][max_capacity]


if __name__ == "__main__":
    weights = [1, 2, 3, 8, 7, 4]
    values = [20, 5, 10, 40, 15, 25]
    capacity = 10
    best = knapsack_solver(weights, values, capacity)
    print("Best possible value:", best)
